ffi/src/engine_data.rs

```diff
@@ -93,3 +93,25 @@ fn get_raw_arrow_data_impl(data: Box<dyn EngineData>) -> DeltaResult<*mut ArrowF
     let ret_data = Box::new(ArrowFFIData { array, schema });
     Ok(Box::leak(ret_data))
 }
+
+#[cfg(feature = "default-engine")]
+#[no_mangle]
+pub unsafe extern "C" fn get_engine_data(
+    data: *mut ArrowFFIData,
+    engine: Handle<SharedExternEngine>,
+) -> ExternResult<Handle<ExclusiveEngineData>> {
+    get_engine_data_impl(data)
+        .map(|engine_data| engine_data.into())
+        .into_extern_result(&engine.as_ref())
+}
+
+#[cfg(feature = "default-engine")]
+unsafe fn get_engine_data_impl(data: *mut ArrowFFIData) -> DeltaResult<Box<dyn EngineData>> {
+    let array_local = std::ptr::read(&(*data).array);
+    let schema_local = std::ptr::read(&(*data).schema);
+
+    let array: StructArray = from_ffi(array_local, &schema_local)?.into();
+    let record_batch: delta_kernel::arrow::array::RecordBatch = array.into();
+    let engine_data = delta_kernel::engine::arrow_data::ArrowEngineData::from(record_batch);
+    Ok(Box::new(engine_data))
```

Comment:
If we change this function to return `ExclusiveEngineData`, can simply do:
```suggestion
    Ok(Box::new(engine_data).into())
```

Reply:
I'm not sure i understood this correctly? I now have:


```rust
#[cfg(feature = "default-engine")]
unsafe fn get_engine_data_impl(
    array: FFI_ArrowArray,
    schema: &FFI_ArrowSchema,
) -> DeltaResult<Handle<ExclusiveEngineData>> {
    let array_data = unsafe { from_ffi(array, schema) };
    let array = StructArray::from(array_data?);
    let record_batch = delta_kernel::arrow::array::RecordBatch::from(array);
    let engine_data: Box<dyn EngineData> = Box::new(
        delta_kernel::engine::arrow_data::ArrowEngineData::from(record_batch),
    );
    Ok(engine_data.into())
}
```